### PR TITLE
Importing Service Provider Configuration from remote Github repository

### PR DESCRIPTION
**Pull request by kaveenr**
### Importing Service Provider Configuration from remote Github repository

- for https://github.com/wso2/carbon-identity-framework/issues/1700
- Git repository manager
- Periodical polling action handler 
- Registry service to register different components to the solution 
- Configuration revision tracking system

### When should this PR be merged

- after https://github.com/wso2/carbon-identity-framework/pull/1737 is merged and rebased
### Follow up actions

- Working on Git authentication methods 
- UI configuration 
- Cache backing DAO's 

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/pom.xml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                        <Export-Package>
+                            org.wso2.carbon.identity.remotefetch.common.*
+                        </Export-Package>
+                    </instructions>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+</project>
```
Comment (madurangasiriwardena) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/pom.xml]: Add new line at the end of the file

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/RemoteFetchConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    private int remoteFetchConfigurationId = -1;
+    private int tenantId = 0;
+    private String repositoryManagerType = "";
+    private String actionListenerType = "";
+    private String confgiurationDeployerType = "";
+    private Map<String, String> repositoryManagerAttributes = new HashMap<>();
+    private Map<String, String> actionListenerAttributes = new HashMap<>();
+    private Map<String, String> confgiurationDeployerAttributes = new HashMap<>();
+
+    public RemoteFetchConfiguration() { // default implementation ignored
```
Comment (madurangasiriwardena) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/RemoteFetchConfiguration.java]: If you don't want the default constructor, you can remove it.

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/RemoteFetchConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                ", repositoryManagerAttributes=" + repositoryManagerAttributes +
+                ", actionListenerAttributes=" + actionListenerAttributes +
+                ", confgiurationDeployerAttributes=" + confgiurationDeployerAttributes +
+                '}';
+    }
+
+    @Override
+    public boolean equals(Object o) {
+
+        if (this == o) return true;
```
Comment (madurangasiriwardena) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/RemoteFetchConfiguration.java]: use below syntax to make it more readable
```
if (this == o) {
    return true;
}
```

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/actionlistener/ActionListenerBuilder.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Builds a ActionListener from a RemoteFetchConfiguration file
+ */
+public abstract class ActionListenerBuilder {
+
+    protected RemoteFetchConfiguration fetchConfig;
+    protected RepositoryManager repoConnector;
+    protected ConfigDeployer configDeployer;
+
+    public ActionListenerBuilder() {
```
Comment (madurangasiriwardena) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/src/main/java/org/wso2/carbon/identity/remotefetch/common/actionlistener/ActionListenerBuilder.java]: As this is the only constructor and it is the default one, you can remove it

File: components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-authentication-flow.jsp [hunk 5]
```diff
@@ -259,7 +259,7 @@
                             <a class="step_order_header" href="#">Step <%=step.getStepOrder()%>
                             </a>
                             <a href="#" class="delete_step icon-link"
-                               style="background-image: url(images/delete.gif);float:right;width: 9px;"></a>
+                               style="background-image: url(images/delete.gif);float:right;width: 9px;" data-step-no="<%=step.getStepOrder()%>"></a>
```
Comment (senthalan) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-authentication-flow.jsp]: Is this change related to this PR?

Comment (senthalan) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-authentication-flow.jsp]: Is this change related to this PR? There are some other files as well, Please check.

Comment (senthalan) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-authentication-flow.jsp]: Is this change related to this PR? There are some other files as well, Please check.

Comment (kaveenr) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-authentication-flow.jsp]: It seems like the branch `wso2:feature-sp-from-vcs` is not rebased with the latest master as the diff between my fork branch and `wso2:master` doesn't contain that file.

(https://github.com/wso2/carbon-identity-framework/compare/master...kaveenr:remotefetch-feature)

Comment (senthalan) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt.ui/src/main/resources/web/application/configure-authentication-flow.jsp]: When you are rebasing with the wso2:master, you can skip other comments.

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/RemoteFetchCore.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Retrieves RemoteFetchConfigurations and builds ActionListeners to be executed
+ */
+public class RemoteFetchCore implements Runnable {
+
+    private Log log = LogFactory.getLog(RemoteFetchCore.class);
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/RemoteFetchCore.java]: private static final

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/RemoteFetchCore.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        ActionListener actionListener;
+        ConfigDeployer configDeployer;
+
+        try {
+            RepositoryManagerBuilder repositoryManagerBuilder = this.componentRegistry
+                    .getRepositoryManagerComponent(fetchConfig.getRepositoryManagerType())
+                    .getRepositoryManagerBuilder();
+
+            repoConnector = repositoryManagerBuilder.addRemoteFetchConfig(fetchConfig).build();
+        } catch (NullPointerException e) {
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/RemoteFetchCore.java]: Use null check instead

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/CacheBackedRemoteFetchConfigurationDAOImpl.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+import org.wso2.carbon.identity.remotefetch.core.cache.RemoteFetchConfigurationByIdCache;
+import org.wso2.carbon.identity.remotefetch.core.cache.RemoteFetchConfigurationByIdCacheEntry;
+import org.wso2.carbon.identity.remotefetch.core.cache.RemoteFetchConfigurationByIdCacheKey;
+import org.wso2.carbon.identity.remotefetch.core.dao.RemoteFetchConfigurationDAO;
+
+import java.util.List;
+
+public class CacheBackedRemoteFetchConfigurationDAOImpl implements RemoteFetchConfigurationDAO {
+
+    RemoteFetchConfigurationDAO remoteFetchConfigurationDAO = new RemoteFetchConfigurationDAOImpl();
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/CacheBackedRemoteFetchConfigurationDAOImpl.java]: private

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/DeploymentRevisionDAOImpl.java [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+    /**
+     * @param deploymentRevision
+     * @return
+     * @throws RemoteFetchCoreException
+     */
+    @Override
+    public int createDeploymentRevision(DeploymentRevision deploymentRevision) throws RemoteFetchCoreException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection();
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/dao/impl/DeploymentRevisionDAOImpl.java]: use JdbcTemplate

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/implementations/actionHandlers/PollingActionListener.java [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+import java.io.File;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PollingActionListener implements ActionListener {
+
+    private static Log log = LogFactory.getLog(GitRepositoryManager.class);
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/implementations/actionHandlers/PollingActionListener.java]: + final

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/implementations/actionHandlers/PollingActionListener.java [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+     * @param deployer
+     */
+    private void pollDirectory(File path, ConfigDeployer deployer) {
+
+        log.info("Polling Directory " + path.getPath() + " for changes");
+        List<File> configFiles = null;
+
+        try {
+            configFiles = this.repo.listFiles(path);
+        } catch (Exception e) {
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/implementations/actionHandlers/PollingActionListener.java]: try to catch specific Exception

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/implementations/actionHandlers/PollingActionListener.java [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+            String currentHash = deploymentRevision.getFileHash();
+
+            try {
+                newHash = this.repo.getRevisionHash(deploymentRevision.getFile());
+            } catch (Exception e) {
+                log.info("Unable to get new hash", e);
+            }
+
+            // Deploy if new file or updated file
+            if (!newHash.isEmpty() && (currentHash.isEmpty() || !(currentHash.equals(newHash)))) {
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/implementations/actionHandlers/PollingActionListener.java]: can change to isFileRevisionChanged(..)

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/implementations/repositoryHandlers/GitRepositoryManager.java [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+    private RevCommit getLastCommit(File path) throws GitAPIException {
+
+        List<RevCommit> revCommits = new ArrayList<>();
+        Iterable<RevCommit> logIterater = git.log().addPath(path.getPath()).call();
+        logIterater.forEach(revCommits::add);
+        return revCommits.get(0);
+    }
+
+    private String inputStreamToString(InputStream inputStream) throws IOException {
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.core/src/main/java/org/wso2/carbon/identity/remotefetch/core/implementations/repositoryHandlers/GitRepositoryManager.java]: Better to use Stream instead of String due to memory pressure.

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/pom.xml [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+                            org.apache.commons.lang; version="${commons-lang.wso2.osgi.version.range}",
+                            org.apache.commons.logging; version="${commons-logging.osgi.version.range}",
+                            org.apache.commons.io; version="${commons.io.wso2.osgi.version.range}",
+                            org.apache.commons.collections; version="${commons-collections.wso2.osgi.version.range}",
+
+                            org.apache.axis2.*; version="${axis2.osgi.version.range}",
+                            org.apache.axiom.*; version="${axiom.osgi.version.range}",
+                        </Import-Package>
+                        <Export-Package>
+                            org.wso2.carbon.identity.remotefetch.common.*
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/pom.xml]: exclude private package

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/pom.xml [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+
+    <artifactId>org.wso2.carbon.identity.remotefetch.common</artifactId>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.wso2.eclipse.osgi</groupId>
+            <artifactId>org.eclipse.osgi.services</artifactId>
```
Comment (ruwanta) [inline components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/pom.xml]: <scope>provided</scope>

File: components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/pom.xml (hunk 14 above)
Comment (kaveenr) [reply components/remotefetch-core/org.wso2.carbon.identity.remotefetch.common/pom.xml]: Common package doesn't contain an internal package, excluded in core
